### PR TITLE
FEATURE: `Neos.Fusion:Component` `@private` props

### PR DESCRIPTION
**Pull request by mhsdesign**
The key `@private` is added to `Neos.Fusion:Component` (and derived prototypes like `Neos.Neos:ContentComponent`) to allow calculating of values for the renderer with access to the props. The `@private` key can only be defined together with the prototype and must not be overridden from outside. Inside the renderer the calculated private values are accessible as private.XXX

This feature replaces various patterns like nested components with `@apply.props` or using the `this` reference.

See the documentation for more details: https://neos.readthedocs.io/en/stable/References/NeosFusionReference.html#neos-fusion-component

**Example::**

```
prototype(Vendor.Site:Component) < prototype(Neos.Fusion:Component) {
    title = 'Hello World'
    titleTagName = 'h1'
    bold = false

    @private {
        computedTitle = ${String.toLowercase(props.title)}
        funnyTitle = Neos.Fusion:Value {
            value = ${props.titleTagName + " " + private.computedTitle}
        }
    }

    renderer = Neos.Fusion:Tag {
        attributes.class {
            component = 'component'
            bold = ${props.bold && 'component--bold'}
        }
        content = Neos.Fusion:Join {
            headline = Neos.Fusion:Tag {
                tagName = ${props.titleTagName}
                content = ${private.funnyTitle}
            }
            // nestedComponentScope = Neos.Fusion:Component {
            //   prop1 = ${props.title} // works
            //   renderer = ${props.title} // doest work!
            // }
      }
    }
}
```


**Review instructions**

|                    |                                                                                                                                                                                                                           |
|--------------------|---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| **[key]**          | (mixed) The public API of your component: Lazy evaluated props that will be available inside the current component's scope under the context ``props`` (is iterable)                                                      |
| **`@private.[key]`** | (mixed) Can only be set inside the root component declaration: Lazy evaluated private props that will be available inside the current component's scope under the context ``private`` (is not iterable / is only a proxy) |
| **renderer**       | (mixed, **required**) The value which gets rendered                                                                                                                                                                       |


> **Note**
> The context ``props`` and ``private`` is only available in the components scope
> The component's scope will be available inside the `renderer` and `@private` and will extend inwards until inside another component's renderer
> That means inside `@private` it's even allowed to reference another private prop (be carefully of circular references, though!)
> But normal props are not inside the component's scope and thus cannot reference each other or ``private``


Spec:
- private is not Iterable but a simple proxy (no var_dump)
- self referencing is allowed
- in scope of the current component (a new component will have as usual its own scope)
- can only be declared within the root prototype declaration (not "at call time")

**Checklist**

- [x] Code follows the PSR-2 coding style
- [x] Tests have been created, run and adjusted as needed
- [x] The PR is created against the [lowest maintained branch](https://www.neos.io/features/release-roadmap.html)
- [x] Reviewer - PR Title is brief but complete and starts with `FEATURE|TASK|BUGFIX`
- [x] Reviewer - The first section explains the change briefly for change-logs
- [x] Reviewer - Breaking Changes are marked with `!!!` and have upgrade-instructions

### CODE REVIEW

**Comment by mhsdesign** [comment]
### Original Draft:
I present to you the pattern of `@private` props:
(similar to  `@context`, one cannot reference one computed prop in another computed prop, but `${props}` is available)
```
prototype(Vendor.Site:Example) < prototype(Neos.Fusion:Component) {
  text = "foo"
  @private {
    constant = 'A long string who is used multiple times, like TailwindCSS classes'
    foo = Neos.Fusion:Value {
      value = ${props.text + "!"}
    }
  }
  renderer = ${props.text + private.foo + private.constant}
}
```

it is strongly inspired by the known and already used patterns shown below https://github.com/neos/neos-development-collection/pull/3943#issuecomment-1298259920



This idea makes https://github.com/neos/neos-development-collection/issues/3702 obsolete
Related discussion: https://discuss.neos.io/t/rfc-fusion-components-simpler-variable-flow-computed-properties-and-public-component-api/5943

**Upgrade instructions**



**Review instructions**



Spec:
- private is not Iterable but a simple proxy (no var_dump)
- self referencing
- In scope of the above component (a new component will have its own scope)

**Checklist**

- [x] Code follows the PSR-2 coding style
- [x] Tests have been created, run and adjusted as needed
- [x] The PR is created against the [lowest maintained branch](https://www.neos.io/features/release-roadmap.html)
- [x] Reviewer - PR Title is brief but complete and starts with `FEATURE|TASK|BUGFIX`
- [x] Reviewer - The first section explains the change briefly for change-logs
- [x] Reviewer - Breaking Changes are marked with `!!!` and have upgrade-instructions



---------
----------


For reference: other computed props pattern in fusion:


### WrapperComponent / NestedComponent pattern:
(most famous i guess)
https://gist.github.com/mficzel/5c9c8e54f5a327123f07d762b34e2053

```
prototype(Vendor.Site:Example) < prototype(Neos.Fusion:Component) {
  foo = null
  bar = null
  renderer = Neos.Fusion:Component {
    @apply.props = ${props}
    barTransformed = Neos.Fusion:Map {
      items = ${props.bar}
      itemRenderer = ...
    }
    
    renderer = afx`{props.bar}  {props.barTransformed}`
  }
}
```

### `renderer.@context` pattern:

i could not find the original "author" but this comes close:
https://gist.github.com/mficzel/5c9c8e54f5a327123f07d762b34e2053?permalink_comment_id=3238803#gistcomment-3238803
and https://discuss.neos.io/t/alternative-to-renderer-context-private/5473

```
prototype(Vendor.Site:Example) < prototype(Neos.Fusion:Component) {
  text = "foo"
  renderer.@context.private = Neos.Fusion:DataStructure {
    foo = ${props.text + "!"}
  }
  renderer = Neos.Fusion:Value {
    // `private` will not be available directly inside of an eel expression on renderer
    value = ${props.text + private.foo}
  }
}
```

### `this` "private" pattern
- limitation: cannot be used for nested computed props

```
prototype(Vendor.Site:Example) < prototype(Neos.Fusion:Component) {
  text = "foo"
   // internal prop by convention
  _foo = ${this.text + "!"}
  renderer = ${props.text + props._foo}
}
```


### `this` `@process` "private" pattern
by @jonnitto 
```
prototype(Vendor.Site:Example) < prototype(Neos.Fusion:Component) {
  text = "foo"
   // internal prop by convention
  _foo = ${this.text}
  _foo.@process.compute = Neos.Fusion:Tag {
    content = ${value}
  }
  renderer = ${props.text + props._foo}
}
```

File: Neos.Fusion/Classes/FusionObjects/ComponentImplementation.php [hunk 1]
```diff
@@ -71,6 +72,17 @@ protected function getProps(array $context): \ArrayAccess
         return $props;
     }
 
+    protected function getComputedProps(array $context): ?\ArrayAccess
+    {
+        $this->runtime->pushContextArray($context);
+
+        $computed = $this->runtime->evaluate($this->path . '/__meta/computed<Neos.Fusion:ComponentComputedProps>');
```
Comment (mhsdesign) [inline Neos.Fusion/Classes/FusionObjects/ComponentImplementation.php]: we need to play ping pong here and ask our friend `Neos.Fusion:ComponentComputedProps` to fetch all the `@computed` props, as the runtime doesnt give us access in the parent component about the subkeys

**Comment by jonnitto** [comment]
Generally, I like this idea. We have to think about the naming. Alternatives to `@computed` would be `@private` or `@internal`

**Comment by mhsdesign** [comment]
Okay lets ...
# Vote

icon  |  choice
------- | --------
🎉 | `@computed`
❤️  | `@private`
🚀| `@internal`
😄| `@protected`


@jonnitto and me decided against `renderer.@computed`, so the meta property should be first level.

or just comment any other proposal

**Comment by markuspfeifenberger** [comment]
When I get this right, the name elected above will get a "protected" word and can not be used as props name any more. If so, wouldn't it make sense, to prefix it with eg `_` ?

**Comment by mhsdesign** [comment]
> When I get this right, the name elected above will get a "protected" word and can not be used as props name any more.

No ;) To prevent this i decided to use a meta property prefixed with `@`. Those cannot be used currently via `props` and are only for special cases like this ;)

**Comment by markuspfeifenberger** [comment]
> > When I get this right, the name elected above will get a "protected" word and can not be used as props name any more.
> 
> No ;) To prevent this i decided to use a meta property prefixed with `@`. Those cannot be used currently via `props` and are only for special cases like this ;)

;-) true ... so maybe `@protected` fits as name as well?

**Comment by jonnitto** [comment]
I know `@protected` is closer to the PHP function type, but: It should be easily understandable, even for a non-PHP developer. That's why I prefer names who describe the goal of this function

**Comment by jonnitto** [comment]
`afx` is more a language similar to `jsx`, so Fusion (and Eel also) is more Front-End focused. There are also in JavaScript classes private constants and functions. And I think it is important to remember that the developer recognizes that these computed props or constants (like a long class name) are not editable. Of course he can do that, but he should not

**Comment by davidspiola** [comment]
IMO @private fits best because it's not changeable from outside and, therefore, side effect free. Also, PHP and JS developers can relate to that pattern. Private **computed** variables, which are much better than nesting the component, just need good documentation. I guess folks will love it :-)

**Comment by mhsdesign** [comment]
> it's not changeable from outside

`prototype(Vendor.Site:Bar).@private.path = "foo bar"`

would still work, but youre right: the api is clearly marked

**Comment by freesh** [comment]
I think technically it's more like protected, but with `@private` it's more understandable in what context it's used.
So i would love to see` @private`. :)

**Comment by mhsdesign** [comment]
one thing bothers me:

we are about to introduce a meta path `@private` which is completely the right choice  but i would be wondering why its not `@renderer` too.

....

Hmm i think i just found my peace with this definition:

the path that is used for the output is a normal path e.g. `value` `renderer` `type`...

a special meta path like `@glue` or `@sortKeys` and `@private` is used to change behaviour of the output path, when all normal arbitary keys are used and api (like in Join, DataStructure or any AbstractArrayFusionObject)

**Comment by mficzel** [comment]
Also think a `@private` matches best as it makes clear that this should not be messed with from outside. Not a huge fan of adding meta properties everywhere but seems the right thing to me as it should perform better than nested components. I also like that you can still access `props` as usual alongside `private`

**Comment by mhsdesign** [comment]
If youre wondering about the breakiness of this change -> its non breaking as when `@private` is not used - the context will not be changed.

There may however be conflicts when `@context.private` and `@private` are used side by side:

```
prototype(Vendor.Site:Example) < prototype(Neos.Fusion:Component) {
  @private {
    foo = "new"
  }
  renderer.@context.private = Neos.Fusion:DataStructure {
    foo = "old"
  }
  renderer = ${private.foo} // will be "old"
}

prototype(Vendor.Site:Example) < prototype(Neos.Fusion:Component) {
  @context.private = Neos.Fusion:DataStructure {
    foo = "old"
  }
  @private {
    foo = "new"
  }
  renderer = ${private.foo} // will be "new"
}
```

i was first trying to throw an exception, which would only be able to forbid the lower - more unlikely case.
The above `renderer.@context.private` is so eager since it sits on renderer, that it overrides the context and there is nothing we can do about it (we cant even detect that this will be happening.)

**Comment by mhsdesign** [comment]
Adjusted to `@private`, cleaned some things up and is now ready for review ;)

File: Neos.Fusion/Resources/Private/Fusion/Prototypes/Component.fusion [hunk 2]
```diff
@@ -1,5 +1,12 @@
 prototype(Neos.Fusion:Component) {
   @class = 'Neos\\Fusion\\FusionObjects\\ComponentImplementation'
   @sortProperties = false
+  @private {
```
Comment (mhsdesign) [inline Neos.Fusion/Resources/Private/Fusion/Prototypes/Component.fusion]: i decided against:

```suggestion
  @private = Neos.Fusion:Internal.PrivateComponentProps {
```

as this makes the fusionobject more api.

File: Neos.Fusion/Classes/FusionObjects/Internal/PrivateComponentPropsImplementation.php [hunk 3]
```diff
@@ -0,0 +1,33 @@
+<?php
+namespace Neos\Fusion\FusionObjects\Internal;
```
Comment (mhsdesign) [inline Neos.Fusion/Classes/FusionObjects/Internal/PrivateComponentPropsImplementation.php]: new namespace for internal things.

i choose this, so when using IDEs to write fusion, one wont accidentally choose `Neos.Fusion:Internal.PrivateComponentProps`

so im against:

- `Neos.Fusion:PrivateComponentProps`
- `Neos.Fusion:Helpers.PrivateComponentProps`

**Comment by kitsunet** [comment]
I was a bit skeptical when I saw this, and think there is still a question about nested computed (eg. I need to compute something from props and then compute sometihng else based on that result again), also naming as Bastian suggested, BUT I am warming to this concept, it seems straight forward, the meta property takes it out of the regular flow and it's pretty easy.

**Comment by mhsdesign** [comment]
> there is still a question about nested computed (eg. I need to compute something from props and then compute sometihng else based on that result again)

youre right - but this is also very hacky with (n) nested components:
https://fusionpen.punkt.de/fusionpen/96a62890723248649f6b130bea4b06575b87d3c0.html

```
prototype(YourVendor:FusionPrototype) < prototype(Neos.Fusion:Component) {
    input = "hi"
    renderer = Neos.Fusion:Component {
        @apply.props = ${props}
        computed1 = ${props.input + '!'}
        renderer = Neos.Fusion:Component {
            @apply.props = ${props}
            computed2 = ${props.computed1 + '!'}
            renderer = ${props.computed2}
        }
    }
}
```

one could do it like this:
```
prototype(YourVendor:FusionPrototype) < prototype(Neos.Fusion:Component) {
    input = "hi"
    @private {
        computed2 = Neos.Fusion:Component {
            // @apply.props = ${props} only if you need the original props in renderer
            computed1 = ${props.input + '!'}
            renderer = ${props.computed1 + '!'}
        }
    }
    renderer = ${private.computed2}
}
```

**Review by jonnitto — APPROVED** [review]
I can't really evaluate the PHP code, but from the functionality I give it a yes.

File: Neos.Fusion/Classes/FusionObjects/ComponentImplementation.php [hunk 4]
```diff
@@ -71,6 +75,17 @@ protected function getProps(array $context): \ArrayAccess
         return $props;
     }
 
+    protected function getPrivateProps(array $context): ?\ArrayAccess
```
Comment (mficzel) [inline Neos.Fusion/Classes/FusionObjects/ComponentImplementation.php]: maybe `getPrivates` ... `getPrivateProps` sounds a bit weird

Comment (kitsunet) [reply Neos.Fusion/Classes/FusionObjects/ComponentImplementation.php]: getPrivates could also be weird though

Comment (mhsdesign) [reply Neos.Fusion/Classes/FusionObjects/ComponentImplementation.php]: and make it `private` :P

i like the current name

Comment (bwaidelich) [reply Neos.Fusion/Classes/FusionObjects/ComponentImplementation.php]: Please not [getPrivates](https://www.linguee.de/deutsch-englisch/search?source=englisch&query=privates) 😄

**Review by grebaldi — APPROVED** [review]
Hi @mhsdesign,

thanks for this implementation :rocket: 

Code LGTM and I was able to verify manually that it works as expected. 

It's a pity that the `private` semantics are not statically enforced atm, but there's probably a way to achieve that at the Parser-level later on. Now, thinking of it, that would be my only remaining question:

Do we agree that static enforcement of `@private` (so that it cannot be overridden at the call-site) is desirable as a future addition to this API? Strictly speaking, we'd have to anticipate that as a future breaking change and at least point out the implication in the documentation to this feature.

On the other hand: If it's not desirable, then `@private` might be a misleading name (I know I'm late to the party regarding this argument :sweat_smile:).

I'm probably overthinking this :smile: So, don't take that question as blocker to this feature. I'm just leaving it here for everyone to consider.

**Comment by mhsdesign** [comment]
@grebaldi excellent suggestion and note! (I would have written that too ^^)

Ill try to tell the AstNodeVisitor or runtime (which may in my opinion hold such dark logic as its not the parser :D) to forbid `@private` outside of the prototype declaration.

If i cant get it done easily (readable) i think we need to satistfy ourselves with a convention ... but we all know what happends to conventions ...

**Comment by manuelmeister** [comment]
I just want to clarify: `@private` only works with Fusion and not with `afx`, because in `afx` we have no way of defining new prototypes, correct?

I have another question: As a frontend dev, I often need to transform data based on multiple props and also use this prop in a computed prop. For instance does this work?

```php
prototype(YourVendor:Counter) < prototype(Neos.Fusion:Component) {

    class = "p-4 bg-black"

    // other possible variants: dense & big
    variant = "default"

    complexObjectInputProp = Neos.Fusion:DataStructure {
        value = 1337
    }

    @private {
        bigOverride = ${props.variant == 'big' || props.complexObjectInputProp.value > 999}
        content = ${Format.number(props.complexObjectInputProp.value)}
        attributes = Neos.Fusion:DataStructure {
            class = ${Array.join(['counter', props.class, private.bigOverride && 'counter--big' ], ' ')}
            data-value = ${props.complexObjectInputProp.value}
        }
    }

    renderer = afx`
       <div {...private.attributes}>
          <YourVendor:Icon @if={private.bigOverride} variant="big" />
          {private.content}
       </div>
    `
}
```

I would like this notation, as I can clearly define the "input" props. Everything inside `@private` is protected. So for example here you can't disable the big override from the outside.

**Comment by mhsdesign** [comment]
hi @manuelmeister thanks for your input.

the above would work 70% as defined by spec:
> (similar to @context, one cannot reference one computed prop in another computed prop, but ${props} is available)

so line 16, referencing `private.bigOverride` would fail. You cannot use private inside of private (at the moment ...)

But maybe it shouldnt be that way - we also allow referencing `this` in another path (`this`).

@kitsunet stated to why `@context` doesnt allow referencing other newly created context items: https://neos-project.slack.com/archives/C050C8FEK/p1668600658634739

> because it's dangerous to rely on the order

> it's separated by design and should stay so


im not sure where to go from know. On the one side time is not on our side (8.2 is near) 
and your point and https://github.com/neos/neos-development-collection/pull/3943#issuecomment-1322045717 could need some work to work with all edge cases ...
i plan on backporting this feature either way as via a package to Neos 7.3 so we might as well wait for 8.3 ...

-------

back to the above - when i wanted to rewrite this to a working approach i noted, its not even really possible, and one needs to step back to other ugly approaches. Thats why im sure this is the way. 

what do you say? Allow referencing private inside private?
Yes: 👍 (Im in for it as this is a huge plus. And its doable easily without any problems as far as i know.)
No: 👎
```
@private {
   bar = "123"
   foo = ${private.bar}
   hello = Neos.Fusion:Tag {
      content = ${private.foo}
  }
}
```

A side advantage would be, that we would disallow any accidental access of a parents `private` props.

Technical implementation:
The lazy private props would use the context with:
`$this->effectiveContext["private"] = $this`

**Comment by manuelmeister** [comment]
I think, within `@private` it would make sense to have something like `this`, but not object local, but `@private` local. This is such a big pain point as a frontend neos-beginner. I just need something that is like `this`, but something that "actually works". With this, component thinking will be much easier.

Countless times I've wanted to assign something and on top of that use this further, but my only option was the `Neos.Fusion:Component` nesting, `@context` or some other workaround. I don't like workarounds because they make the code hard to read.

**Comment by kitsunet** [comment]
I think technically we could also remove the constraint for `@context`. A big problem with all of this (also `this` as it works currently, pun intended) is that you can create indirect infinite loops and I find that very unfortunate. Too often in the past have I seen people create recursive self references between variables. I don't think that's great and i am also not sure we can find a way to prevent this really (other than not allowing references like that in the first place)

**Comment by mhsdesign** [comment]
yes infinite loops are a point (i just thought one should avoid them, like bugs - you want to avoid them too :P)
But ill give you that its odd to be able to completely destroy your site by writing some self referencing (declarative) fusion - which is only found out at runtime ... but as this is already possible via this, and some odd prototype inheritance we might add yet another way :D

the best thing we can do is detect an infinite loop and throw an exception (Thats what i will do).

**Comment by kitsunet** [comment]
The problem will be detecting infinite loops, but I guess we can see when we get to that...

**Comment by mhsdesign** [comment]
i have a plan for that already ;)

basically a map of paths that are currently beeing evaluated, and if it tries to recurse - it will be noticed.

**Comment by kitsunet** [comment]
> 

Yeah but a) that is probably expensive to keep track of and b) recursion is quite common in Fusion (up to a point), and keeping track of like hundredths of steps deep stacks, is def expensive... But lets see

**Comment by mhsdesign** [comment]
@kitsunet solved only for this case (not for `this`) without hurting performance see 61f2c47bf8ab4ca07de191de80280d21fd3d7bd4

**Comment by mhsdesign** [comment]
This has been already approved 2 times, but id like to get a third review (or a revisit from @grebaldi) about 61f2c47bf8ab4ca07de191de80280d21fd3d7bd4

File: Neos.Fusion/Classes/FusionObjects/ComponentImplementation.php [hunk 5]
```diff
@@ -55,6 +57,10 @@ public function evaluate()
     protected function prepare(array $context): array
     {
         $context['props'] = $this->getProps($context);
+        $private = $this->getPrivateProps($context);
+        if ($private !== null) {
+            $context['private'] = $private;
```
Comment (mhsdesign) [inline Neos.Fusion/Classes/FusionObjects/ComponentImplementation.php]: we should consider always setting the context `private` (at leasz to null) so there is a consistent behavior that outer private props will never pass.

or maybe less breaking: check if `private` is instance of lazyProps and disallow only that

**Comment by mficzel** [comment]
The last commit adds quite a bit. The whole concept of internal fusion prototypes is new. Would it do harm to merge the state before that. 

The last commit deserves a separate pr

**Comment by mficzel** [comment]
Other things to adjust:

1. Add documentation for @private to `Neos.Neos/Documentation/References/NeosFusionReference.rst` 
2. Add more descriptive pr message for release notes. Can be adjusted later but before release.

**Review by mficzel — CHANGES_REQUESTED** [review]
The additional fusion prototypes to render props using props should be added in a separate pr as it is a distinct feature. The state before the last commit would be fine with me.

Also:
- Add documentation for `@private` to Neos.Neos/Documentation/References/NeosFusionReference.rst ... can be one line
- Add more descriptive pr message for release notes. Can be adjusted later but before release.

**Comment by mhsdesign** [comment]
@mficzel thank you for your evaluation

@Sebobo told me, that he also wanted to review and try this out - so im definitely delaying this feature for 8.3

**Comment by mhsdesign** [comment]
~Sadly there is a catch: The following wont work:~
edit: of course it doesnt work and its by design:

```
component.privateScopeIsIsolatedInRenderer = Neos.Fusion:Component {
  @private {
    outer = "Will not be reachable"
  }
  renderer = Neos.Fusion:Component {
    renderer = Neos.Fusion:Component {
      prop1 = ${private.outer}
      renderer = ${props.prop1}
    }
  }
}
```

The second `Neos.Fusion:Component` acts as a barrier and wont let outer private props pass ... (The outer private context is not merged but always replaced for isolation purposes)

for comparison, this will work:

```
component.privateScopeIsReachableInProps = Neos.Fusion:Component {
  @private {
    outer = "Moin"
  }
  renderer = Neos.Fusion:Component {
    prop1 = ${private.outer}
    renderer = ${props.prop1}
  }
}
```

**Comment by mhsdesign** [comment]
As discussed with @mficzel we could clean up the code and removed a hack, by dropping the feature to `@apply` or loop over private props.
This is imo a wise choice.

Simply said as the test prove it you can neitger use `@apply.it = ${private}` nor `Array.join(private)` since it is no longer iterable but just a proxy

good catch from @manuelmeister -> you wont be able to debug private props ... (currently)

File: Neos.Fusion/Classes/FusionObjects/Helpers/LazySelfReferentialProps.php [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+            $this->runtime->pushContextArray($this->effectiveContext);
+            try {
+                $this->valueCache[$path] = $this->runtime->evaluate($this->parentPath . '/' . $path);
+            } finally {
+                $this->runtime->popContext();
+                unset($this->currentlyEvaluatingIndexes[$path]);
+            }
+        }
+        return $this->valueCache[$path];
+    }
```
Comment (grebaldi) [inline Neos.Fusion/Classes/FusionObjects/Helpers/LazySelfReferentialProps.php]: I'd like to suggest a slightly different mechanism to keep the responsibility of circular reference checking out of `LazySelfReferentialProps`.

First, I'd like to propose a dedicated class that encapsulates a(ny) lazy value reference:
```php
/**
 * @template T
 */
#[Flow\Proxy(false)]
final class LazyReference
{
    private bool $isLocked = false;
    private bool $hasBeenDereferenced = false;

    /**
     * @var T
     */
    private mixed $value = null;

    /**
     * @param callable():T $calculateValue
     */
    public function __construct(
        private readonly mixed $calculateValueCallback
    ) {
    }

    /**
     * @return T
     */
    public function deref(): mixed
    {
        if ($this->hasBeenDereferenced) {
            return $this->value;
        }

        if ($this->isLocked) {
            throw new \Exception('Circular reference!');
        }

        $this->isLocked = true;
        $this->value = $this->calculateValue();
        $this->hasBeenDereferenced = true;

        return $this->value;
    }

    /**
     * @return T
     */
    private function calculateValue(): mixed
    {
        $calculateValueCallback = $this->calculateValueCallback;
        return $calculateValueCallback();
    }
}
```

`LazyReference`'s `deref` method takes care of circular reference checking by locking itself upon evaluation. This mechanism can easily tested via unit test:

**Case 1: No dependencies**
```php
$lazyRef = new LazyReference(function () {
    return 'foo';
});
$lazyRef->deref(); // returns "foo"
```

**Case 2: Non-circular dependencies**
```php
$lazyRef1 = new LazyReference(function () use (&$lazyRef2) {
    return $lazyRef2->deref();
});
$lazyRef2 = new LazyReference(function () use (&$lazyRef3) {
    return $lazyRef3->deref();
});
$lazyRef3 = new LazyReference(function () {
    return 'foo';
});

$lazyRef1->deref(); // returns "foo"
```

**Case 3: Self-dependent**
```php
$lazyRef = new LazyReference(function () use (&$lazyRef) {
    return $lazyRef->deref();
});
$lazyRef->deref(); // throws "Circular reference!"
```

**Case 4: Transitive, circular dependencies**
```php
$lazyRef1 = new LazyReference(function () use (&$lazyRef2) {
    return $lazyRef2->deref();
});
$lazyRef2 = new LazyReference(function () use (&$lazyRef3) {
    return $lazyRef3->deref();
});
$lazyRef3 = new LazyReference(function () use (&$lazyRef1) {
    return $lazyRef1->deref();
});

$lazyRef1->deref(); // throws "Circular reference!"
```

*Side Note: Of course, `LazyReference` can be set up to carry an identifier, so it is able to throw a more helpful error message*

Now, with `LazyReference` in place, the `offsetGet` method of `LazySelfReferentialProps` can be implemented like this:
```php
public function offsetGet($path): mixed
{
    $fullyQualifiedPath = $this->parentPath . '/' . $path;

    $this->lazyReferences[$path] ??= new LazyReference(
        function () use ($fullyQualifiedPath) {
            $this->runtime->pushContextArray($this->effectiveContext);
            try {
                return $this->runtime->evaluate($fullyQualifiedPath);
            } finally {
                $this->runtime->popContext();
            }
        }
    );

    return $this->lazyReferences[$path]->deref();
}
```

Wdyt, @mhsdesign?

Comment (mhsdesign) [reply Neos.Fusion/Classes/FusionObjects/Helpers/LazySelfReferentialProps.php]: this is most certainly the most beautifull solution thanks for you proposal ;D

**Comment by manuelmeister** [comment]
@mhsdesign I'm confused. Does this mean that the following wont work:

## Definitions
```jsx
prototype(Your.Vendor:Counter) < prototype(Neos.Fusion:Component) {
  isGlobal = false
  count = 0

  @private {
    icon = ${props.isGlobal ? "world" : "flag"}
  }

  renderer = afx`
    <Your.Vendor:Button class="m-counter">
       <Your.Vendor:Icon name={private.icon} />
       Counter: {props.count}
    </Your.Vendor:Button>
  `
}

prototype(Your.Vendor:Button) < prototype(Neos.Fusion:Component) {
  class = null
  content = null

  @private {
      classes = Neos.Fusion:DataStructure {
         identifier = 'm-btn'
         additionalClasses = ${props.class}
      }
  }

  renderer = afx`
    <button class={private.classes}>
      {props.content}
    </button>
  `
}

prototype(Your.Vendor:Icon) < prototype(Neos.Fusion:Component) {
  name = "edit"

  renderer = afx`
    <svg @if={props.name == 'flag'} class="svg-flag">…</svg>
    <svg @if={props.name == 'world'} class="svg-world">…</svg>
    <svg @if={props.name == 'edit'} class="svg-edit">…</svg>
  `
}
```

## Usage
```jsx
<Your.Vendor:Counter isGlobal count={3} />
```

## Result
```html
<button class="m-btn m-counter">
  <svg class="svg-world">…</svg>
  Counter: 3
</button>
```

**Comment by markusguenther** [comment]
Here we also need to rebase on 8.3 and change the target.

**Comment by mhsdesign** [comment]
@manuelmeister I can reassure you ;) it's exactly what you expected! It works ;)

In https://github.com/neos/neos-development-collection/pull/3943#issuecomment-1332612988 i ment something else what i think will never be an issue and is more "by desing"
Its acutally the same behavior with props. They are not available in another components renderer.
But in your example we pass the private prop in a non renderer path "content" which is okay ;)

File: Neos.Fusion/Classes/FusionObjects/Helpers/LazySelfReferentialProps.php (hunk 6 above)
Comment (mhsdesign) [reply Neos.Fusion/Classes/FusionObjects/Helpers/LazySelfReferentialProps.php]: implemented ;)

**Comment by mhsdesign** [comment]
Lets go ;)

`@private` is now truly worthy of its name ^^

image

```
# works:
prototype(Your.Vendor:Counter) < prototype(Neos.Fusion:Component) {
    @private {
        bar = "foo"
    }
}

# forbidden:
hihi = Your.Vendor:Counter {
    @private {
        bar = "foo"
    }
}

# also forbidden:
prototype(Your.Vendor:Counter) {
    @private {
        bar = "foo"
    }
}
```

**Comment by mhsdesign** [comment]
Added docs ;)

can be proof read here: https://github.com/neos/neos-development-collection/blob/feature/fusionComponentEasyComputedProps/Neos.Neos/Documentation/References/NeosFusionReference.rst#neosfusioncomponent

**Comment by mhsdesign** [comment]
@mficzel its now finally ready to roll out ;)

**Comment by mhsdesign** [comment]
Do we want to take encapsulation even further?

we could allow an alternate renderer path inside of `@private` … just for thought …
This won’t be able to be overridden from outside then…

edit: i reserved `@private.renderer` for the future ;) https://github.com/neos/neos-development-collection/pull/3943/commits/a72f00b9d00aa85df0e32b7fa767901e67aa3051

File: Neos.Fusion/Classes/FusionObjects/Helpers/LazySelfReferentialProps.php [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+
+    public function offsetGet($path): mixed
+    {
+        $fullyQualifiedPath = $this->parentPath . '/' . $path;
+
+        $this->lazyReferences[$path] ??= new LazyReference(
+            function () use ($fullyQualifiedPath) {
+                $this->runtime->pushContextArray($this->effectiveContext);
+                try {
+                    return $this->runtime->evaluate($fullyQualifiedPath);
```
Comment (mhsdesign) [inline Neos.Fusion/Classes/FusionObjects/Helpers/LazySelfReferentialProps.php:66]: we can make it fail hard if we want to rely on the private prop to exist which sounds nice ...

Comment (manuelmeister) [reply Neos.Fusion/Classes/FusionObjects/Helpers/LazySelfReferentialProps.php:66]: What do you mean?

Comment (mhsdesign) [reply Neos.Fusion/Classes/FusionObjects/Helpers/LazySelfReferentialProps.php:66]: That you cannot do `private.nonExistant` but an exception will be thrown ;)

Edit, is now implemented https://github.com/neos/neos-development-collection/pull/3943/commits/18e0515dc72270508b8a581d7c33ca489586c632

**Pull request merged**